### PR TITLE
TEP-0114: Add support for PipelineRun reconciler to create CustomRuns

### PR DESCRIPTION
**Pull request by abayer**
# Changes

Modifies the `PipelineRun` reconciler to allow creating `CustomRun`s, rather than `Run`s, if the new `custom-task-version` feature flag is set to `v1beta1`, rather than the default value of `v1alpha1`.

See [TEP-0114](https://github.com/tektoncd/community/blob/main/teps/0114-custom-tasks-beta.md#new-feature-flag-custom-task-version) for more information on the feature flag.

The `Run`-flavored `wait-task` is renamed to `wait-task-alpha` for better clarity.

Also adds a new integration test for using the beta `CustomRun` with the `custom-task-version` feature flag set to `v1beta1`.

Followups needed:
* Docs updates
* Deprecation notice on the `custom-task-version` flag
* After the next release, `test/custom-task-ctrls/wait-task-beta/cmd/controller/main.go` needs to switch to using `tkncontroller.FilterCustomRunRef`
* And also after that release, `test/custom-task-ctrls/wait-task-beta/pkg/reconciler/reconciler_test.go` needs to switch to using `parse.MustParseCustomRun`

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Custom task runs created from PipelineRuns can now be v1beta1.CustomRuns instead of v1alpha1.Runs, if the "custom-task-version" feature flag is set to "v1beta1", instead of the default "v1alpha1". Note that custom task controllers would need to be updated to listen for *v1beta1.CustomRun instead of *v1alpha1.Run in order to respond to v1beta1.CustomRuns.
```

### CODE REVIEW

File: test/custom-task-ctrls/wait-task-beta/cmd/controller/main.go
Comment (abayer) [inline test/custom-task-ctrls/wait-task-beta/cmd/controller/main.go:53]: The relevant changes have been merged (https://github.com/tektoncd/pipeline/pull/5822) but unless we change the `wait-task-beta`'s `go.mod` to depend on a specific sha of `tektoncd/pipeline` instead of a release, we still need to wait for the next to release to switch this properly.

File: pkg/apis/config/feature_flags.go
Comment (XinruZhang) [inline pkg/apis/config/feature_flags.go:86]: Based on [TEP-0114 New Feature Flag `custom-task-version`](https://github.com/tektoncd/community/blob/main/teps/0114-custom-tasks-beta.md#new-feature-flag-custom-task-version)
> We set the default feature flag value to `v1alpha1` in the first LTS release, and switch to `v1beta1` in the following three LTS releases.

We may want to set the default value to `v1alpha1` in the upcoming release `v0.43.0`, switch the default value to `v1beta1` in the release `v0.44.0`, and delete the feature flag in the release `v0.47.0`

File: test/custom-task-ctrls/wait-task-beta/cmd/controller/main.go
Comment (XinruZhang) [inline test/custom-task-ctrls/wait-task-beta/cmd/controller/main.go:59]: non-CustomRun

File: test/custom-task-ctrls/wait-task-beta/cmd/controller/main.go
Comment (XinruZhang) [inline test/custom-task-ctrls/wait-task-beta/cmd/controller/main.go:63]: Wanna confirm if my understanding is correct:

End users are allowed to define a CustomRun without `CustomRef` and `CustomSpec` specified right? For example: https://github.com/tektoncd/experimental/blob/main/wait-task/example-run.yaml

**Review by XinruZhang — COMMENTED** [review]
Thanks @abayer!

File: docs/install.md
Comment (abayer) [reply docs/install.md:430]: 👍

File: pkg/apis/config/feature_flags.go
Comment (abayer) [reply pkg/apis/config/feature_flags.go:86]: Ok.

Comment (abayer) [reply pkg/apis/config/feature_flags.go:86]: I'll get to this later today, 'cos I have to rewrite a _lot_ of tests. =)

Comment (XinruZhang) [reply pkg/apis/config/feature_flags.go:86]: NP~ Thank you! ❤️

**Comment by abayer** [comment]
/hold

@XinruZhang - default changed. I'm holding the PR for the moment so that if for some reason we change our minds and want `v1beta1` to be the default again, I don't have to spend an hour or two undoing what I just did. =)

**Review by XinruZhang — COMMENTED** [review]
Several minor comments, thanks!

**Comment by jerop** [comment]
> I'm holding the PR for the moment so that if for some reason we change our minds and want `v1beta1` to be the default again, I don't have to spend an hour or two undoing what I just did. =)

I support defaulting to `v1alpha1` in the initial release as described in [TEP-0114](https://github.com/tektoncd/community/blob/main/teps/0114-custom-tasks-beta.md#new-feature-flag-custom-task-version)

cc @tektoncd/core-maintainers

**Comment by abayer** [comment]
/hold cancel

Ok, that's enough confirmation for me - I've squashed and removed the hold.

File: docs/install.md
Comment (jerop) [inline docs/install.md:436]: tried to clarify the documentation for the feature flag:

```suggestion
- `custom-task-version`: set this flag to "v1alpha1" to have `PipelineRuns` create `Runs` from Custom Tasks. Set it to "v1beta1" to have `PipelineRuns` create the beta `CustomRuns`. This may be needed if you are using Custom Tasks that have been updated to listen for `*v1beta1.CustomRun` instead of `*v1alpha1.Run`. For more information, see [Runs](runs.md) and [CustomRuns](customruns.md). Flag defaults to "v1alpha1". 

```

File: pkg/apis/config/feature_flags_test.go
Comment (jerop) [inline pkg/apis/config/feature_flags_test.go:137]:
```suggestion
				CustomTaskVersion:               config.DefaultCustomTaskVersion,
```

File: pkg/apis/config/feature_flags.go
Comment (jerop) [inline pkg/apis/config/feature_flags.go:250]: let's add a test for an invalid value in `TestNewFeatureFlagsConfigMapErrors` https://github.com/tektoncd/pipeline/blob/f62563fadf809f5def179a8f40ae3dcfc6316c7f/pkg/apis/config/feature_flags_test.go#L193-L212

File: pkg/apis/pipeline/v1alpha1/run_types.go
Comment (jerop) [inline pkg/apis/pipeline/v1alpha1/run_types.go:270]: let's add unit tests for this function

File: pkg/apis/pipeline/v1beta1/customrun_types.go
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/customrun_types.go:263]: let's add unit tests for this function

File: pkg/apis/pipeline/v1beta1/pipelinerun_types.go
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/pipelinerun_types.go:529]:
```suggestion
	// Status is the CustomRunStatus for the corresponding CustomRun
```

Comment (abayer) [reply pkg/apis/pipeline/v1beta1/pipelinerun_types.go:529]: Actually, these should all be "CustomRun or Run", since `PipelineRunRunStatus` now applies to both. Fixing!

File: pkg/reconciler/pipelinerun/cancel.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/cancel.go:227]: can we use the constant for "v1alpha1" instead?

Comment (abayer) [reply pkg/reconciler/pipelinerun/cancel.go:227]: ha, missed that, fixing.

File: config/config-feature-flags.yaml
Comment (XinruZhang) [inline config/config-feature-flags.yaml:78]: Should we switch the default value to `true` in v0.43.0 release or v0.44.0 release?

Comment (XinruZhang) [reply config/config-feature-flags.yaml:78]: cc @jerop

Comment (jerop) [reply config/config-feature-flags.yaml:78]: v0.43, we have an issue tracking this - https://github.com/tektoncd/pipeline/issues/5155

File: pkg/reconciler/pipelinerun/cancel_test.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/cancel_test.go:294]:
```suggestion
						TypeMeta:         runtime.TypeMeta{Kind: pipeline.RunControllerName},
```

Comment (jerop) [reply pkg/reconciler/pipelinerun/cancel_test.go:294]: also, can we please add `customRuns` to the `TestCancelPipelineRun` test?

Comment (abayer) [reply pkg/reconciler/pipelinerun/cancel_test.go]: Done!

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:68]: Renaming this to `RunObjectName` will make it clearer that it's either Run or CustomRun name

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:69]: Renaming this to `RunObject` will make it clearer that it's either Run or CustomRun

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:70]: Renaming this to `RunObjectNames` will make it clearer that it's either Run or CustomRun names

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:71]: Renaming this to `RunObjects` will make it clearer that it's either Runs or CustomRuns

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/pipelinerun.go:944]:
```suggestion
func (c *Reconciler) createRunObjects(ctx context.Context, rpt *resources.ResolvedPipelineTask, pr *v1beta1.PipelineRun) ([]v1beta1.RunObject, error) {
```

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/pipelinerun.go:945]:
```suggestion
	var runObjects []v1beta1.RunObject
```

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/pipelinerun.go:958]:
```suggestion
func (c *Reconciler) createRunObject(ctx context.Context, runName string, params []v1beta1.Param, rpt *resources.ResolvedPipelineTask, pr *v1beta1.PipelineRun) (v1beta1.RunObject, error) {
```

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/pipelinerun.go:1364]:
```suggestion
func updatePipelineRunStatusFromChildObjects(ctx context.Context, logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, taskRuns []*v1beta1.TaskRun, runObjects []v1beta1.RunObject) error {
```

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/pipelinerun.go:1530]:
```suggestion
func updatePipelineRunStatusFromCustomRunsOrRuns(logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, runObjects []v1beta1.RunObject) {
```

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/pipelinerun.go:1557]:
```suggestion
func updatePipelineRunStatusFromChildRefs(logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, trs []*v1beta1.TaskRun, runObjects []v1beta1.RunObject) {
```

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/pipelinerun.go:316]:
```suggestion
		getRunObjectFunc := func(name string) (v1beta1.RunObject, error) {
```

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/pipelinerun.go:331]:
```suggestion
			getRunObjectFunc = func(name string) (v1beta1.RunObject, error) {
```

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/pipelinerun.go:350]:
```suggestion
			getRunObjectFunc,
```

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/pipelinerun.go:1339]:
```suggestion
	var runObjects []v1beta1.RunObject
```

File: pkg/reconciler/pipelinerun/pipelinerun_test.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/pipelinerun_test.go:4886]: it seems that this change is unrelated 🤔

File: pkg/reconciler/pipelinerun/pipelinerun_test.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/pipelinerun_test.go:5298]: it seems that this change is unrelated 🤔

File: pkg/reconciler/pipelinerun/timeout.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/timeout.go:147]:
```suggestion
	for _, customRunName := range customRunNames {
```

File: pkg/reconciler/pipelinerun/pipelinerun_test.go
Comment (abayer) [reply pkg/reconciler/pipelinerun/pipelinerun_test.go:4886]: Ah, right - so `TestReconcileWithPipelineResults` was calling both of `runTestReconcileWithFinallyResults` _and_ `runTestReconcileWithPipelineResults`. I had to fix some testcases/code in those, and it was really confusing as to what exactly was failing, so while I was here, I split them up to actually make sense. Unrelated, yeah, but might as well do it here.

File: pkg/reconciler/pipelinerun/pipelinerun_test.go
Comment (abayer) [reply pkg/reconciler/pipelinerun/pipelinerun_test.go:5298]: See above.

File: pkg/apis/run/v1beta1/customrunstatus_types.go
Comment (abayer) [reply pkg/apis/run/v1beta1/customrunstatus_types.go:151]: Done.

File: pkg/reconciler/pipelinerun/pipelinerun_test.go
Comment (jerop) [reply pkg/reconciler/pipelinerun/pipelinerun_test.go:4886]: I see, thanks for the clarification!

**Review by jerop — APPROVED** [review]
thank you @abayer!

**Comment by dibyom** [comment]
Thanks @abayer - this looks good to me!

There are quite a few follow ups - are we tracking them somewhere besides the PR description (just want to make sure we get all of them in for the release)

**Comment by dibyom** [comment]
/lgtm

/hold

(feel free to cancel the hold if the docs/deprecation notice stuff is already being tracked and is on track for the release)

**Comment by jerop** [comment]
> looks like we only need 2 of the follow ups before the release

Documentation: tracking docs in https://github.com/tektoncd/pipeline/issues/5158 and there's a PR for migration docs https://github.com/tektoncd/pipeline/pull/5850

Deprecation notice on the `custom-task-version` flag: tracking in https://github.com/tektoncd/pipeline/issues/5836 and https://github.com/tektoncd/pipeline/issues/5837

/hold cancel

**Pull request merged**
